src/shared/sam/cli/samCliDetection.ts

```diff
@@ -16,34 +16,41 @@ const localize = nls.loadMessageBundle()
 const lock = new AsyncLock()
 
 const learnMore = localize('AWS.samcli.userChoice.visit.install.url', 'Get SAM CLI')
-
 const browseToSamCli = localize('AWS.samcli.userChoice.browse', 'Locate SAM CLI...')
-
 const settingsUpdated = localize('AWS.samcli.detect.settings.updated', 'Settings updated.')
-
 const settingsNotUpdated = localize('AWS.samcli.detect.settings.not.updated', 'No settings changes necessary.')
 
-export async function detectSamCli(showMessageIfDetected: boolean): Promise<void> {
+/** Most-recent resolved SAM CLI location. */
+let currentsamCliLocation: string | undefined = undefined
+
+/**
+ *
+ * @param opt.showMessage true: always show message, false: never show
```

Comment:
This is great but the implementation is a bit confusing, would it be better to just have some strings as the type? Something like `{ showMessage: 'always' | 'never' | 'diff' }` (or even just `'always' | diff'` and have the param as optional)?

Reply:
> Something like `{ showMessage: 'always' | 'never' | 'diff' }`

I did that originally, but opted for the keyword-args style. 

-  since the caller must specify `showMessage`, and the docstring specifies the semantics, I think it's reasonably clear.
- the ad-hoc `always|never|...` values looked obtuse/overwrought in the impl
- the `boolean|undefined` is more readily recognized as a "tri-state" for readers